### ISSUE TITLE
Node cannot recover from unexpected shutdown

### ISSUE THREAD

**Issue by dkijania** [input-output-hk/jormungandr#2592]
**Describe the bug**
In rare occasions we can observe that node looses data on abrupt restart. In simple scenario we bootstrap node with some genesis block, send one simple transaction and then node is killed.

From time to time we observe that node after restart does not have in history mentioned transaction.

**Mandatory Information**
1. `jcli --full-version`: newest master
2. `jormungandr --full-version` newest master

**To Reproduce**
1. Go to jormungandr_integration_tests
2. Run test `jormungandr::recovery::test_node_recovers_kill_signal`
3. Observe error

**Expected behavior**
Node should recover from kill signal with 100% correctness

**Additional context**
All test data: [tmpvZ8H5X.zip](https://github.com/input-output-hk/jormungandr/files/5367007/tmpvZ8H5X.zip)
command to run node
```
jormungandr.exe  --secret  node_secret.yaml --config node_config.yaml --genesis-block block-0.bin


```

**Comment by eugene-babichenko** [comment]
This is a storage problem. After the restart the tip can be one block behind. In the logs you provided:

- Here is the message before the restart: `{"msg":"update current branch tip: 454c129a-00000004-0.4 -> 97db7c29-00000005-0.5","level":"INFO","ts":"2020-10-12T17:11:05.017689+02:00","date":"0.5","parent":"454c129a3628d063df58218fdd79963129c19e15032381c69223587f3b45f0d1","hash":"97db7c297e763904031922bda86467ac27a2a15c3da8e1f93bab942924d0e384","task":"block"}`
- But after the restart we have: `{"msg":"Loaded from storage tip is : 454c129a-00000004-0.4","level":"INFO","ts":"2020-10-12T17:11:10.805142+02:00","task":"bootstrap"}`

This is because sled does only periodic flushes triggered by its internal timer. This is why this bug can be hard to reproduce: it is very dependent on the timing which is actually a bit of a randomized factor.

The solution could be to forcibly flush the storage after updating the tag, but this needs to be investigated. I am not sure if the `fsync` system call (or its Windows equivalent) would finish upon receiving `SIGKILL`. If not, this can lead to new problems.

**Comment by eugene-babichenko** [comment]
Since `sled` does its own caching, there is still a possibility that the `.flush()` call to the database will be interrupted by `SIGKILL`. Thus, even if we force flushing, the test may still fail on rare occasions. Although doing more frequent flushes reduces the possibility of such failures.

**Comment by dkijania** [comment]
This looks like another issue of single node network. In standard cases node would try to update its ledger from the network. So probably the best thing to do is alter the test, so it can allow situation in which abruptly shutdown single node can loose recent  data.

**Comment by mzabaluev** [comment]
> the best thing to do is alter the test, so it can allow situation in which abruptly shutdown single node can loose recent data.

I agree. The important thing that we _do_ want to test for is whether the node survives an abrupt shutdown and is able to catch up on any new blocks (including a block that the node itself may have recently created and propagated to other nodes). Loss of transactions in the block that was not yet flushed should not be treated as a failure.

"Regular" tests should not use `SIGKILL` to shut down and restart a node.

**Comment by Zeegomo** [comment]
seems to be fixed after #2921
